rfcs/0045-zkapp-balance-data-in-archive.md

```diff
@@ -0,0 +1,253 @@
+  token_permissions   token_permissions  NOT NULL
+  balance             bigint             NOT NULL
+  nonce               bigint             NOT NULL
+  receipt_chain_hash  text               NOT NULL
+  delegate            int                          REFERENCES public_keys(id)
+  voting_for          text    NOT NULL
+  timing              int                          REFERENCES timing_info(id)
+  permissions         int     NOT NULL             REFERENCES zkapp_account(id)
+  zkapp               int                          REFERENCES snapp_account
+  zkapp_uri           text    NOT NULL
+```
+
+The new type `token_permissions` is:
+
+  `CREATE TYPE token_permission AS ENUM ('token_owned_new_accounts_disabled', 'token_owned_new_accounts_enabled',
+                                         'token_not_owned_account_disabled', 'token_not_owned_account_enabled')`
+
+(Alternatively, we could create a new table with four entries corresponding to these possibilities,
+and refer to it with a foreign key.)
+
+The new table `tokens` is:
+```
+  id                  serial  PRIMARY KEY
+  token_id            text    NOT NULL
+  token_symbol        text    NOT NULL
+```
+
+Table `public_keys`, add column:
+
+  `ledger_index int  NOT NULL`
+
+In order to include the hard fork genesis ledger accounts in this table, we may need
+a separate app to populate it. Alternatively, once we have the genesis ledger, we
+could use an app to dump the SQL needed to populate the table, and keep that SQL in the
+Mina repository.
+
+Delete the unused table `zkapp_party_balances`.
+
+### Changes to the daemon-to-archive-process RPC
+
+Blocks are sent from the daemon to the archive process via an RPC named `Send_archive_diff`.
+There are several kinds of messages that can be sent using that RPC, but the one
+we're interested in here is the `Breadcrumb_added` message, which contains a block.
+
+Currently, the per-transaction balances (for user commands and internal commands only, not for zkApps)
+are contained in the transaction statuses of transactions contained in a block. The
+transactions are contained in the `Staged_ledger_diff.t` part of the block, which
+is an instance of the type `External_transition.t`.
+
+We'll still want to send blocks using this RPC, but remove some information, and add some
+other information.
+
+The RPC type is unversioned, because the daemon and archive process should come from the
+same build. Hence, the changes proposed here don't require adding versioning.
+
+Information to be removed:
+
+- In `Transaction.Status.t`, the `Applied` constructor is applied to a pair consisting of
+  `Auxiliary_data.t` containing account creation fees, and `Balance_data.t`, containing
+  balances. The fields in those types are all options, because they may not be relevant
+  to a particular transaction. Make the constructor unary, applied to a new type
+  `Account_creation_fees.t` (see below).
```

Comment:
I think we can extract this information from the scan state (in staged ledger). The scan state stores `Transaction_applied.t` which for every transaction type (except zkapps, for now) has `previous_empty_accounts`. All we have to do it find the transactions added as part of the current block.

Reply:
I had a look, and I think we can add the fees in the same places where we create `internal_command_balances` (which won't be needed, I think, at least not for the archive db).

Reply:
But it would still be part of blocks, no? I was suggesting removing it from blocks altogether and getting it from staged ledger's scan state and RPCed to the archive node